### PR TITLE
8252999: Cleanup: replace .equals("") with .isEmpty() within all codebase

### PR DESCRIPTION
**Pull request by doom369**
**isEmpty** is faster + has less byte code + easier to read. Benchmarks could be found [here](https://medium.com/javarevisited/micro-optimizations-in-java-string-equals-22be19fd8416).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8252999](https://bugs.openjdk.java.net/browse/JDK-8252999): Cleanup: Replace .equals("") with .isEmpty() throughout the codebase ⚠️ Title mismatch between PR and JBS.

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/29/head:pull/29`
`$ git checkout pull/29`

### CODE REVIEW

**Comment by thatsIch** [comment]
@doom369 jcheck requires an associated issue

File: src/demo/share/java2d/J2DBench/src/j2dbench/tests/iio/InputImageTests.java
Comment (mrserb) [inline src/demo/share/java2d/J2DBench/src/j2dbench/tests/iio/InputImageTests.java:187]: This file intentionally maintains compatibility to jdk1.4, so equals("") should be used.

File: src/demo/share/java2d/J2DBench/src/j2dbench/tests/iio/OutputImageTests.java
Comment (mrserb) [inline src/demo/share/java2d/J2DBench/src/j2dbench/tests/iio/OutputImageTests.java:146]: This file intentionally maintains compatibility to jdk1.4, so equals("") should be used.

**Comment by doom369** [comment]
@mrserb hello. Thanks for the review. Any further actions required from me?

**Comment by kevinrushforth** [comment]
Before this Enhancement can be formally reviewed, you will need a JBS bug ID. If you are already working with a Committer or Reviewer in the `jdk` project who has agreed to sponsor your change, they can file the Enhancement request. Otherwise, you can file it using the web interface at [bugreport.java.com](https://bugreport.java.com/). Once you have a JBS bug ID, you need to edit the PR title to include that bug ID (without the `JDK-`) replacing "Improvement". 

Since this PR cuts across many functional areas (each gray label represents a functional area), you should expect a longer review process, since someone from each functional area will need to look at the changes in their area (like @mrserb started to do for the `2d` area).

**Comment by doom369** [comment]
@kevinrushforth thanks. Done.

**Comment by AlanBateman** [comment]
The code in java.base was updated to use String::isEmpty in JDK 12 (JDK-8215281). There was follow-up in JDK 13 to do the same in the java.desktop module (JDK-8223237). Changing the remaining usages make sense although I see that more more than half are in tests.

It would be good to hear from security-dev on the changes to the Apache Santuario code (in java.xml.crypto module) in case it would be better to contribute those upstream instead. Ditto for the Apache Xalan code (in the java.xml module) but it may be significantly forked already that it doesn't matter.

I assume you can use JDK-8215014 rather than creating a new issue.

**Comment by dholmes-ora** [comment]
This should be broken up to deal with the files in different functional areas under different bugids and PRs. Otherwise the cross-posting to so many lists is prohibitive. Files in different areas need to be reviewed by different teams. Thank you.

**Comment by doom369** [comment]
I have in mind dozens of improvements all over the code like this one. I already did some further review and in most cases, those tiny changes go trough all codebase. I can create PR for every separate module, but in general, it would bring like multiply x5 the number of PRs in total. If you think it's a better way to go, no problem, I can do that.

**Comment by kevinrushforth** [comment]
@dholmes-ora raises a good point. Hopefully there is a balance point between a dozen different bugs / pull requests each targeting one area and one bug / PR targeting a dozen separate areas. I don't have a good general rule to suggest. Maybe @AlanBateman or @jddarcy can offer some advice?

**Comment by dholmes-ora** [comment]
14 cc'd mailing lists is unworkable. You need to be subscribed to lists to post, but even if you are a reply-all will be delayed due to all of the mails being held up for moderator approval due to:

" Too many recipients to the message"

I have a longer email coming once it gets through the moderator approval delay. :(

**Comment by JornVernee** [comment]
Since one of the motivations for this change is micro benchmark performance, please add a benchmark to the JDKs micro benchmark suite as well. (see e.g. https://github.com/openjdk/jdk/tree/master/test/micro/org/openjdk/bench/java/lang).

Also, what testing has been done for these changes?

**Comment by jonathan-gibbons** [comment]
> I have in mind dozens of improvements all over the code like this one. 

That sounds scary. Broad updates like these cause unnecessary churn in the codebase, and can make merging and back porting harder.  Changes should be discussed ahead of time with the appropriate teams.

**Review by prrace — CHANGES_REQUESTED** [review]
1) This is un-necessary churn.
2) I can't even be sure I am finding the ones in my area because there's so much here
3) The ones I can find have no need of whatever performance improvement this might bring.
I think this whole PR should be withdrawn, and there may an attempt at measuring the benefits of this for the various cases before submitting in appropriate smaller chunks. But I'll tell you now I don't see a need for the test updates you are making.

**Comment by doom369** [comment]
Ok, sorry for the distraction.

**Comment by bradfordwetmore** [comment]
Our local Santuario maintainer says:

In general, changes to Apache Santuario should also be made at Apache so we stay in sync.

**Comment by stuart-marks** [comment]
Hi @doom369, I hope we didn't end up wasting too much of your time with this. I wanted to respond to a comment you made earlier in this PR,

> I have in mind dozens of improvements all over the code like this one.

It's hard to see, but as you discovered, the JDK has different groups of people maintaining different areas, and sometimes there are hidden constraints on those different areas, for example, to avoid divergence with upstream source bases. And as you discovered, sometimes those source bases might need to maintain compatibility with an older JDK ... so we don't want to update this code even though it's IN the JDK.

Those kind of constraints generally don't apply to code in the java.base module, since there's nothing upstream of it, and by definition it cannot depend on anything outside the JDK. Generally -- though there are exceptions -- we're more receptive to keeping stuff in java.base (and sometimes related modules close to the core) on the latest and greatest stuff. There are some constraints, however. There are some things we can't use too early during initialization of the JDK, such as lambdas. Also, there is some code lurking around that is sometimes executed by the boot JDK, which is typically one release behind. (This is definitely the case for tools like javac, but I think it might also apply to some things in base.)

Anyway, if you'd like to pursue some of these improvements, drop a note to core-libs-dev@openjdk and we can talk about it.

Thanks.

**Comment by doom369** [comment]
@stuart-marks thanks. Sure, I understand that maintaining OpenJDK is not a simple task. I thought as change is super simple without any side effects it can go through. But I was wrong. That's fine. I'll focus on `java.base` when I have some free time.

**Pull request closed without merging**
